flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java

```diff
@@ -94,26 +130,218 @@ public RegistrationResponse apply(final JobMasterGateway jobMasterGateway) {
-	 * @param taskExecutorAddress      The address of the TaskExecutor that registers
-	 * @param resourceID               The resource ID of the TaskExecutor that registers
+	 * @param resourceManagerLeaderId The fencing token for the ResourceManager leader
+	 * @param taskExecutorAddress     The address of the TaskExecutor that registers
+	 * @param resourceID              The resource ID of the TaskExecutor that registers
 	 *
 	 * @return The response by the ResourceManager.
 	 */
 	@RpcMethod
-	public org.apache.flink.runtime.rpc.registration.RegistrationResponse registerTaskExecutor(
-			UUID resourceManagerLeaderId,
-			String taskExecutorAddress,
-			ResourceID resourceID) {
+	public Future<org.apache.flink.runtime.rpc.registration.RegistrationResponse> registerTaskExecutor(
+		final UUID resourceManagerLeaderId,
+		final String taskExecutorAddress,
+		final ResourceID resourceID
+	) {
+		log.info("received register from taskExecutor {}, address {}", resourceID, taskExecutorAddress);
+		Future<TaskExecutorGateway> taskExecutorFuture =
+			getRpcService().connect(taskExecutorAddress, TaskExecutorGateway.class);
+
+		return taskExecutorFuture.map(new Mapper<TaskExecutorGateway, org.apache.flink.runtime.rpc.registration.RegistrationResponse>() {
+			@Override
+			public org.apache.flink.runtime.rpc.registration.RegistrationResponse apply(final TaskExecutorGateway taskExecutorGateway) {
+				// decline registration if resourceManager cannot connect to the taskExecutor using the given address
+				if(taskExecutorGateway == null) {
+					log.warn("resourceManager {} decline registration from the taskExecutor {}, cannot connect to it using given address {} ",
+						getAddress(), resourceID, taskExecutorAddress);
+					return new org.apache.flink.runtime.rpc.registration.RegistrationResponse.Decline("cannot connect to taskExecutor using given address");
+				} else {
+					// save the register taskExecutor gateway
+					taskExecutorGateways.put(resourceID, taskExecutorGateway);
+					// schedule the heartbeat with the registered taskExecutor
+					ResourceManagerToTaskExecutorHeartbeatScheduler heartbeatScheduler = new ResourceManagerToTaskExecutorHeartbeatScheduler(
+						ResourceManager.this, leaderSessionID, taskExecutorGateway, taskExecutorAddress, resourceID, log);
+					heartbeatScheduler.start();
+					heartbeatSchedulers.put(resourceID, heartbeatScheduler);
+					return new TaskExecutorRegistrationSuccess(new InstanceID(), heartbeatScheduler.getHeartbeatInterval());
+				}
+			}
+		}, getMainThreadExecutionContext());
+
+	}
+
+	/**
+	 * notify resource failure to resourceManager, because of two reasons:
+	 * 1. cannot keep heartbeat with taskManager for several times, mark the resource as failed
+	 * 2. in some corner cases, TM will be marked as invalid by cluster manager master(e.g. yarn master), but TM itself does not realize.
+	 *
+	 * @param resourceID identify the taskManager which to stop
+	 */
+	@RpcMethod
```

Comment:
in some corner cases, Cluster manager framework(e.g. yarn master) will notify resourceManager that a TaskExecutor is marked as invalid, so it could be a RpcMethod